### ISSUE TITLE
TextController.clear() crashes on iOS

### ISSUE THREAD

**Issue by collinjackson** [flutter/flutter#9362]
## Steps to Reproduce

Make a text field with a button that clears it.

```
class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => new _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  TextEditingController _controller = new TextEditingController();

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      body: new Center(
        child: new TextField(controller: _controller)
      ),
      floatingActionButton: new FloatingActionButton(
        onPressed: () => _controller.clear(),
        tooltip: 'Clear',
        child: new Icon(Icons.clear),
      ),
    );
  }
}
```

Type some text into the text field and press the clear button. It crashes on iOS simulator.

## Logs

```
Apr 12 21:17:10 jackson-macbookpro Runner[52288]: *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: 'The index 1 is invalid'
	*** First throw call stack:
	(
		0   CoreFoundation                      0x0000000111630b0b __exceptionPreprocess + 171
		1   libobjc.A.dylib                     0x0000000111095141 objc_exception_throw + 48
		2   CoreFoundation                      0x0000000111699625 +[NSException raise:format:] + 197
		3   Foundation                          0x0000000110bc4e20 -[NSString rangeOfComposedCharacterSequenceAtIndex:] + 73
		4   Flutter                             0x000000010f309846 -[FlutterTextInputView positionFromPosition:offset:] + 187
		5   UIKit                               0x0000000112059ad1 -[UIResponder(UITextInput_Internal) _characterInRelationToRangedSelection:] + 257
		6   UIKit                               0x0000000111d348e5 -[UIKeyboardImpl(UIKitInternal) setInitialDirection] + 160
		7   UIKit                               0x0000000111d1ab19 -[UIKeyboardImpl completeUpdateForChangedSelection:] + 124
		8   UIKit                               0x0000000111d1a91e __64-[UIKeyboardImpl updateForChangedSelectionWithExecutionContext:]_block_invoke_3 + 91
		9   UIKit                               0x00000001125901dd -[UIKeyboardTaskExecutionContext returnExecutionToParentWithInfo:] + 278
		10  UIKit                               0x0000000112590812 -[UIKeyboardTaskQueue continueExecutionOnMainThread] + 374
		11  UIKit                               0x00000001125912e4 -[UIKeyboardTaskQueue waitUntilTaskIsFinished:] + 253
		12  UIKit                               0x0000000112591455 -[UIKeyboardTaskQueue performSingleTask:] + 268
		13  UIKit                               0x0000000111d12ee4 -[UIKeyboardImpl textChanged:] + 122
		14  UIKit                               0x0000000111d1abed -[UIKeyboardImpl selectionDidChange:] + 95
		15  Flutter                             0x000000010f3090c9 -[FlutterTextInputView setTextInputState:] + 224
		16  Flutter                             0x000000010f30a32e -[FlutterTextInputPlugin handleMethodCall:resultReceiver:] + 315
		17  Flutter                             0x000000010f303211 __45-[FlutterMethodChannel setMethodCallHandler:]_block_invoke + 123
		18  Flutter                             0x000000010f31032c _ZN5shell21PlatformMessageRouter21HandlePlatformMessageEN3ftl6RefPtrIN5blink15PlatformMessageEEE + 176
		19  Flutter                             0x000000010f3135b1 _ZN5shell15PlatformViewIOS21HandlePlatformMessageEN3ftl6RefPtrIN5blink15PlatformMessageEEE + 35
		20  Flutter                             0x000000010f332ea1 _ZNSt3__110__function6__funcIZN5shell6Engine21HandlePlatformMessageEN3ftl6RefPtrIN5blink15PlatformMessageEEEE3$_1NS_9allocatorIS9_EEFvvEEclEv + 59
		21  Flutter                             0x000000010f319e53 _ZN3fml15MessageLoopImpl15RunExpiredTasksEv + 1001
		22  Flutter                             0x000000010f31c72e _ZN3fml17MessageLoopDarwin11OnTimerFireEP16__CFRunLoopTimerPS0_ + 26
		23  CoreFoundation                      0x00000001115c4454 __CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__ + 20
		24  CoreFoundation                      0x00000001115c40e3 __CFRunLoopDoTimer + 1011
		25  CoreFoundation                      0x00000001115c3caa __CFRunLoopDoTimers + 266
		26  CoreFoundation                      0x00000001115bb9ec __CFRunLoopRun + 1916
		27  CoreFoundation                      0x00000001115bb016 CFRunLoopRunSpecific + 406
		28  GraphicsServices                    0x0000000115e66a24 GSEventRunModal + 62
		29  UIKit                               0x0000000111a540d4 UIApplicationMain + 159
		30  Runner                              0x000000010f284e5f main + 80
		31  libdyld.dylib                       0x0000000118e3265d start + 1
	)
```

## Flutter Doctor

✓] Flutter (on Mac OS X 10.12.2 16C67, channel unknown)
    • Flutter at /Users/jackson/git/flutter
    • Framework revision 819bb539f1 (5 hours ago), 2017-04-12 16:25:12 -0700
    • Engine revision bcc8779cb8
    • Tools Dart version 1.23.0-dev.11.6

[✓] Host Executable Compatibility
    • Downloaded executables execute on host

[✓] Android toolchain - develop for Android devices (Android SDK 25.0.2)
    • Android SDK at /Users/jackson/Library/Android/sdk/
    • Platform android-25, build-tools 25.0.2
    • ANDROID_HOME = /Users/jackson/Library/Android/sdk/
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version: OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] iOS toolchain - develop for iOS devices (Xcode 8.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 8.3.1, Build version 8E1000a
    • ios-deploy 1.9.1
    • CocoaPods version 1.0.1

[✓] Android Studio (version 2.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Gradle version 3.2
    • Java version: OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] IntelliJ IDEA Community Edition (version 2016.3.4)
    • Dart plugin version 163.13137
    • Flutter plugin version 0.1.11.2

[✓] Connected devices
    • iPhone SE • 3BCD157D-B028-4044-B56D-D1E912AA9811 • ios • iOS 10.3 (simulator)

**Comment by collinjackson** [comment]
git bisect say it broke in @cbracken's engine roll

commit 7b0b5c5760eaa0279782c1fa061b3d7a7acf2ba9
Author: Chris Bracken <chris@bracken.jp>
Date:   Wed Apr 12 14:43:34 2017 -0700

    Roll engine to 059ab3a47165d98cb1d7ac5e2a2e4b50abf39832 (#9352)

**Comment by cbracken** [comment]
Thanks. Will add a framework test that covers this as well.

**Comment by cbracken** [comment]
Fixed in flutter/engine#3590. Rolled in #9382.
